lib/glimesh_web/templates/user_settings/edit.html.eex

```diff
@@ -196,4 +196,95 @@
+                <div class="card-body">
+
+                    <h3>2FA Setup</h3>
+
+                    <%= form_for @tfa_changeset, Routes.user_settings_path(@conn, :update_tfa), fn f -> %>
+                    <%= if @tfa_changeset.action do %>
+                    <div class="alert alert-danger">
+                        <p style="margin-bottom: 0px;">Oops, something went wrong! Please check the errors below.</p>
+                    </div>
+                    <% end %>
+                    <div id="tfa-form-loader" style="display:none;">
+                    
+                    <div class="new-tfa-form-class">
+                        <div class="row">
+                            <div class="col-sm-12">
+                                <div id="tfa-img">
+                                </div>
+                            </div>
+                        </div>
+                        <br />
+                    </div>
+
+                    <div class="row">
+                        <div class="col-sm-6">
+                            <div class="form-group">
+                                <%= label f, "2FA Code" %>
+                                <%= text_input f, :tfa, required: true, class: "form-control mb-4" %>
+                                <%= error_tag f, :tfa %>
+                            </div>
+                        </div>
+                        <div class="new-tfa-form-class col-sm-6">
+                            <div class="form-group">
+                                <%= label f, :current_password, for: "current_password_for_tfa" %>
+                                <%= password_input f, :current_password, required: true, name: "current_password", id: "current_password_for_tfa", class: "form-control mb-4" %>
+                                <%= error_tag f, :current_password %>
+                            </div>
+                        </div>
+                    </div>
+
+                    <div style="display:none;">
+                        <div class="form-group">
+                            <%= label f, :tfa_tmp, for: "tmp_for_tfa" %>
+                            <%= password_input f, :tfa_tmp, required: true, name: "tfa_tmp", id: "tmp_for_tfa", class: "form-control mb-4" %>
+                            <%= error_tag f, :tfa_tmp %>
+                        </div>
+                    </div>
+
+                    <div>
+                        <div class="new-tfa-form-class">
+                            <%= submit "Submit Two Factor Authentication", class: "btn btn-primary" %>
+                        </div>
+                        <div class="existing-tfa-form-class">
+                            <%= submit "Remove Two Factor Authentication", class: "btn btn-primary" %>
+                        </div>
+                    </div>
+                    </div>
+                    <% end %>
+                </div>
+            </div>
+        </div>
+    </div>
+
 </div>
+
+<script>
```

Comment:
What does this code really power? Couldn't we handle this all on the server side to not have this javascript exist?

Reply:
the reason I did this javascript code is that I couldn't find a method to safely store the 2FA hash code between requests and if there is a propper way to do this I can change it to use that method instead. but this is how the system know later what hash to use to check the 2FA codes with.